### PR TITLE
Linked PCL Modules with the external libraries for simplifying transitive dependencies

### LINKED ISSUE #3740: [CMake] question about transitives external dependencies
## Your Environment

* Operating System and version: Linux Mint 19.3
* Compiler: g++-8
* PCL Version: Latest release 1.10.0 and dev branch

## Context

Hello,
I've got a issue/question about how PCL handles external dependencies. I know that you recently change some piece of CMake code but in my opinion there is still a lack of "export" of external libraries (FLANN, VTK, ...) when using PCL COMPONENTS .

## Expected Behavior

When I use PCL in a project I would like to do something like:

```cmake

find_package(PCL 1.10 REQUIRED COMPONENTS common filters features registration search kdtree io)

#Other cmake stuff
... 

target_link_libraries(myProject
    PUBLIC
    pcl_common
    pcl_filters 
    pcl_features 
    pcl_registration 
    pcl_search 
    pcl_kdtree
    pcl_io
    )
```

## Current Behavior

But currently if I do what I explain in _Expected Behavior_ there are plenty of external libraries that are missing.

We don't want to explicitly link `myProject` against FLANN, VTK or whatever since they are only needed by PCL.

So I take a look in the [PCLConfig.cmake.in file](https://github.com/PointCloudLibrary/pcl/blob/master/PCLConfig.cmake.in) and it seems like there is all the needed information when parsing the PCL components [here](https://github.com/PointCloudLibrary/pcl/blob/master/PCLConfig.cmake.in#L332) in the `${PCL_${COMPONENT}_LIBRARIES}` , but this variable is [reset later in the script](https://github.com/PointCloudLibrary/pcl/blob/master/PCLConfig.cmake.in#L659) but never used.

## Possible Solution

What I suggest is to add `${PCL_${COMPONENT}_LIBRARIES}` in the `INTERFACE_LINK_LIBRARIES`. That's what we have done [here](https://github.com/IRCAD-IHU/pcl/commit/7d0fe5f5c26d35128a5e6741ae0e6edf93d595b3), this way PCL can be find using CMake target and external libraries such as FLANN seems to be transitive and hidden to final user.

What do you think ? Did I miss a point ?

### PR DESCRIPTION
**Pull request by schweitzer**
Add external libraries of each PCL component to it's INTERFACE_LINK_LIBRARIES.

This ensure transitivity of external libraries when using PCL module/components. 

Closes #3740

### CODE REVIEW

**Comment by SergioRAgostinho** [comment]
@schweitzer would you mind rebasing on the current master. The ci jobs should be fixed now.

**Comment by schweitzer** [comment]
@SergioRAgostinho sorry I miss your comment, I've rebased this on latest master branch.

**Comment by kunaltyagi** [comment]
Issue in building downstream target "tutorials"

**Comment by taketwo** [comment]
To be honest, I don't like how `PCL_${COMPONENT}_LIBRARES` is first used to collect external dependencies, and then is reset to be just the component target. I propose to change line 332 to append external libraries to `PCL_${COMPONENT}_LINK_LIBRARIES` instead. These lists are already linked to the targets, so no further changes will be required.

The Tutorials CI job uncovered some issues with this PR though:

```
CMake Error at /__w/1/install/share/pcl-1.10/PCLConfig.cmake:683 (set_target_properties):
  Property INTERFACE_LINK_LIBRARIES may not contain link-type keyword
  "optimized".  The INTERFACE_LINK_LIBRARIES property may contain
  configuration-sensitive generator-expressions which may be used to specify
  per-configuration rules.
```

Apparently some of the external libraries have link-type keywords in the list of libraries that they provide. If I understand right, `pcl_remove_duplicate_libraries` macro can filter out these. So a potential fix will be to to pass `PCL_${COMPONENT}_LINK_LIBRARIES` through this macro before using it in setting `INTERFACE_LINK_LIBRARIES` on imported PCL targets.

**Comment by schweitzer** [comment]
@taketwo : Thanks for the review, I've applied your suggestion (appending external libraries to `PCL_${COMPONENT}_LINK_LIBRARIES`).

Concerning the tutorials issue, I've faced the same issue on my side but it seems that someone already try to fix it. If you take a look on [line 654](https://github.com/PointCloudLibrary/pcl/blob/master/PCLConfig.cmake.in#L654) it uses `target_link_library` as a workaround. Since this is only available on recent cmake version we can not replace the remaining ` INTERFACE_LINK_LIBRARIES`.

I think I've try to use `pcl_remove_duplicate_libraries` to remove the keywords but it wasn't working, I need to re-check this part. 

Anyway if `pcl_remove_duplicate_libraries` doesn't remove keywords, we can add a `loop` like this one to filter external libraries before appending them:

```cmake
...
      foreach(lib ${${LIB}_LIBRARIES})
        if("${lib}" STREQUAL "debug")
          set(_waiting_for_debug 1)
        elseif("${lib}" STREQUAL "optimized")
          set(_waiting_for_optimized 1)
        elseif(_waiting_for_debug AND NOT ${CMAKE_BUILD_TYPE} STREQUAL "Debug")
          # Reject library
          set(_waiting_for_debug 0)
        elseif(_waiting_for_optimized AND ${CMAKE_BUILD_TYPE} STREQUAL "Debug")
          # Reject library
          set(_waiting_for_optimized 0)
        else()
          list(APPEND PCL_${COMPONENT}_LINK_LIBRARIES_FILTERED "${lib}")
        endif()
      endforeach()
      list(APPEND PCL_${COMPONENT}_LINK_LIBRARIES "${PCL_${COMPONENT}_LINK_LIBRARIES_FILTERED}")
```

**Comment by taketwo** [comment]
> I think I've try to use pcl_remove_duplicate_libraries to remove the keywords but it wasn't working, I need to re-check this part.

You are right. I just skimmed the file the last time and noticed `pcl_remove_duplicate_libraries` does _something_ about link-type keywords. Not that I had a closer look, I see it does not remove them at all.

Something along the lines of the snippet you posted is the way forward indeed :+1:

**Comment by larshg** [comment]
I think this could lead to troubles for windows/visual studio uses, as its configuring for both debug and release at the same time.
Maybe look into converge them to configuration expressions.
I'll happily verify it before we merge.

**Comment by taketwo** [comment]
Yes, ideally each third-party dependency should be an imported target with properly set per-configuration link libraries. Not every third-party library is well-behaved though, so we'd need to create these imported targets ourselves.

@schweitzer which of the third-party libraries have "optimized"/"debug"?

**Comment by schweitzer** [comment]
Hello, 

@larshg : I didn't think about windows/visual studio uses, indeed it can lead to issues when using it.

I'm not sure about what you mean by:
> Maybe look into converge them to configuration expressions.

@taketwo : As far as I remember we have trouble with FLANN. But in the case of this MR I don't know which one is involved, `Tutorials` pipeline fails with this cmake error:

```
-- Found PCL_SURFACE: /__w/1/install/lib/libpcl_surface.so  
CMake Error at /__w/1/install/share/pcl-1.10/PCLConfig.cmake:683 (set_target_properties):
  Property INTERFACE_LINK_LIBRARIES may not contain link-type keyword
  "optimized".  The INTERFACE_LINK_LIBRARIES property may contain
  configuration-sensitive generator-expressions which may be used to specify
  per-configuration rules.
Call Stack (most recent call first):

```

No further details are given. So I think I need to reproduce the `Tutorials` configuration on my side to check which library is adding the flags.

This can take a while since I'm working remotely (since COVID19) and I don't have the same dev env as at office :wink:

**Comment by larshg** [comment]
I know 3. party libs to vtk has it.
As I recently added this pr to vcpkg https://github.com/microsoft/vcpkg/pull/10518

**Comment by schweitzer** [comment]
I've managed to create the same env as in "Tutorials" CI, in a docker container (Ubuntu 16.04).

I take as an example the `cloud_viewer` tutorial.

I've added log in PCLConfig.cmake  like this:
`message("-- Link Libraries ${PCL_${COMPONENT}_LINK_LIBRARIES}")`

And this is the result when I configure `cloud_viewer` tutorial

```
...
-- looking for PCL_SURFACE
-- Link Libraries: optimized;/usr/lib/x86_64-linux-gnu/libqhull.so;debug;/usr/lib/x86_64-linux-gnu/libqhull.so;pcl_common;pcl_search;pcl_kdtree;pcl_octree
CMake Error at /usr/local/share/pcl-1.10/PCLConfig.cmake:685 (set_target_properties):
  Property INTERFACE_LINK_LIBRARIES may not contain link-type keyword
  "optimized".  The INTERFACE_LINK_LIBRARIES property may contain
  configuration-sensitive generator-expressions which may be used to specify
  per-configuration rules.
Call Stack (most recent call first):
  CMakeLists.txt:5 (find_package)

```
So for this particular case, flags are coming from `libqhull`.

I don't know if this is helpful :sweat_smile:

**Comment by taketwo** [comment]
@schweitzer are you comfortable enough with CMake to revamp our finder script for QHull a little? The idea would be to update this section:

https://github.com/PointCloudLibrary/pcl/blob/ca64747a2c3813280aa9040d5b3ccb6725838915/cmake/Modules/FindQhull.cmake#L85-L94

Create a new imported `qhull` target and setup its properties. You can take the corresponding section in FLANN finder script as an example:

https://github.com/PointCloudLibrary/pcl/blob/ca64747a2c3813280aa9040d5b3ccb6725838915/cmake/Modules/FindFLANN.cmake#L128-L156

**Comment by schweitzer** [comment]
@taketwo : Yes sure I can give a try, I let you know if I need more help :wink:

**Comment by taketwo** [comment]
Sure, feel free to ask for advice and sorry in advance if the answers come with a bit of latency :)

**Comment by schweitzer** [comment]
First attempt of modifying FindQhull.cmake.

I've Created an imported target called `QHULL:QHULL` and modified some CMake files of PCL to use the target instead the old `QHULL_INCLUDE_DIRS` & `QHULL_LIBRARIES`.

Still in `FindQhull.cmake` I've also added a piece of code to try to find Qhull using a modern cmake approach, since it seems that in [latest version of Qhull](https://github.com/qhull/qhull/blob/master/CMakeLists.txt#L621), config files are generated. For now I didn't manage to test this part correctly.

I've tested this successfully by building pcl & cloud_viewer tutorial on 3 docker containers (Ubuntu 16.04, 18.04 & 19.04).

**Comment by schweitzer** [comment]
Seems there is more work to do to make in work on all OS :wink:

**Comment by SergioRAgostinho** [comment]
Our Windows CI is currently broken. Don't rely on it for now.

**Comment by larshg** [comment]
@schweitzer there  might be some difficulties with windows and vcpkg. See this pr https://github.com/microsoft/vcpkg/pull/9923.

So start with linux/macs.

It's been a while since I worked / tested the vcpkg PR.

**Comment by schweitzer** [comment]
Ok, it seems to work on both macOS & Linux now, that's a promising first step !

Concerning windows, I think it is the same error that @larshg point me on vcpkg, unfortunately I didn't have a windows environment to test it, so I may need some help for this.

File: cmake/Modules/FindQhull.cmake
Comment (SergioRAgostinho) [inline cmake/Modules/FindQhull.cmake:54]: Is this important to be set? This is usually a CACHE variable that users usually supply during config stage.

File: cmake/Modules/FindQhull.cmake
Comment (SergioRAgostinho) [inline cmake/Modules/FindQhull.cmake:58]: The version is being unconditionally set to 6. It feels wrong.

File: cmake/Modules/FindQhull.cmake
Comment (SergioRAgostinho) [inline cmake/Modules/FindQhull.cmake:70]: This is still being executed even if the config file is found. Is it required? I would expect most this section to be entirely skipped if the config file is found.

File: cmake/Modules/FindQhull.cmake
Comment (SergioRAgostinho) [inline cmake/Modules/FindQhull.cmake:126]: `--Dqh_QHpointer_dllimport` was the double `-` intentional? Also, does `cl` (windows compiler accept) definitions prefixed with `-D`. I think it's supposed to be `/D`.
https://docs.microsoft.com/en-us/cpp/build/reference/d-preprocessor-definitions?view=vs-2019
Help, windows people @larshg , @Morwenn

File: cmake/Modules/FindQhull.cmake
Comment (SergioRAgostinho) [inline cmake/Modules/FindQhull.cmake:93]: is it not worth to report this?

Comment (Morwenn) [reply cmake/Modules/FindQhull.cmake:93]: It's absolutely worth keeping it: I often need to know where my library has been found and it's a simple way to diagnose errors when users share logs.

File: cmake/Modules/FindQhull.cmake
Comment (Morwenn) [reply cmake/Modules/FindQhull.cmake:126]: I believe that `target_compile_definitions(QHULL::QHULL PUBLIC qh_QHpointer)` is the preferred portable way to achieve this without having to care about compiler-specific syntax.

File: cmake/Modules/FindQhull.cmake
Comment (Morwenn) [reply cmake/Modules/FindQhull.cmake:70]: As I mentioned in the Discord this actually seems to be properly skipped when the config file is found.

File: cmake/Modules/FindQhull.cmake
Comment (taketwo) [inline cmake/Modules/FindQhull.cmake:119]: We are missing this in the "modern config" branch.

File: cmake/Modules/FindQhull.cmake
Comment (schweitzer) [reply cmake/Modules/FindQhull.cmake:58]: Yes, I didn't change this part of the file, it seems that the version is used later as an hint to find the right library name. It seems that he major version is rather **7** than **6** on most OS, but I think we shouldn't rely on a version number to find the library.

I can remove this, and see what happend, what do you think ?

File: cmake/Modules/FindQhull.cmake
Comment (schweitzer) [reply cmake/Modules/FindQhull.cmake:93]: I've removed it because I thought it was a duplicate.

When I configure cloud_viewer (with this PR) I've got this

```
root@367c56bb4824:/opt/pcl/tutorials/cloud_viewer# cmake ../../doc/tutorials/content/sources/cloud_viewer/ -DCMAKE_BUILD_TYPE=Release
-- The C compiler identification is GNU 5.4.0
-- The CXX compiler identification is GNU 5.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Checking for module 'eigen3'
--   Found eigen3, version 3.2.92
-- Found Eigen: /usr/include/eigen3 (Required is at least version "3.1") 
-- Eigen found (include: /usr/include/eigen3, version: 3.2.92)
-- Boost version: 1.58.0
-- Found the following Boost libraries:
--   system
--   filesystem
--   date_time
--   iostreams
--   serialization
--   regex
-- Checking for module 'flann'
--   Found flann, version 1.8.4
-- Found FLANN: /usr/lib/x86_64-linux-gnu/libflann_cpp.so  

...

-- Found Qhull: /usr/lib/x86_64-linux-gnu/libqhull.so  <-----
``` 

I don't know where this message comes from, maybe from one of the cmake find function ?

Anyway if you think the previous message is still needed, I can fix it.

File: cmake/Modules/FindQhull.cmake
Comment (schweitzer) [reply cmake/Modules/FindQhull.cmake:119]: Good catch, I will fix it.

File: cmake/Modules/FindQhull.cmake
Comment (Morwenn) [reply cmake/Modules/FindQhull.cmake:93]: For some reason I don't get the *Found Qhull* at every rebuild, but only sometimes.

Comment (schweitzer) [reply cmake/Modules/FindQhull.cmake:93]: Yep, seems to be printed only at first configure ?

File: cmake/Modules/FindQhull.cmake
Comment (schweitzer) [reply cmake/Modules/FindQhull.cmake:126]: It was indeed a mistake, I will add @Morwenn solution !

File: cmake/Modules/FindQhull.cmake
Comment (Morwenn) [reply cmake/Modules/FindQhull.cmake:93]: Yeah, while the other ones are systematically printed. I don't know why there is a difference.

File: cmake/Modules/FindQhull.cmake
Comment (schweitzer) [reply cmake/Modules/FindQhull.cmake:126]: Actually there is one issue with this:
 `target_compile_definitions`  only support `INTERFACE` items on IMPORTED targets, but this isn't available on old cmake version (like the 3.5.1 on Ubuntu 16.04).

So compiler specific flags may still be needed...

Comment (Morwenn) [reply cmake/Modules/FindQhull.cmake:126]: Oh, that's unfortunate :/

File: cmake/Modules/FindQhull.cmake
Comment (SergioRAgostinho) [reply cmake/Modules/FindQhull.cmake:58]: :+1: Go for it.

File: cmake/Modules/FindQhull.cmake
Comment (schweitzer) [reply cmake/Modules/FindQhull.cmake:126]: Indeed :( , I'm still looking for a better solution.

I don't know how QHULL works but does the definition `qh_QHpointer_dllimport` replace `qh_QHpointer` on MSVC, or both definitions are needed ?

File: cmake/Modules/FindQhull.cmake
Comment (schweitzer) [reply cmake/Modules/FindQhull.cmake:93]: I've re-added the previous message function.

File: cmake/Modules/FindQhull.cmake
Comment (schweitzer) [reply cmake/Modules/FindQhull.cmake:115]: Should be fixed

File: cmake/Modules/FindQhull.cmake
Comment (schweitzer) [reply cmake/Modules/FindQhull.cmake]: I've copied this from the findFLANN.cmake file, this is not used, so I removed it.

File: cmake/Modules/FindQhull.cmake
Comment (schweitzer) [reply cmake/Modules/FindQhull.cmake]: Removed !

File: cmake/Modules/FindQhull.cmake
Comment (schweitzer) [reply cmake/Modules/FindQhull.cmake]: For now, I've kept the `if(MSVC)`, and updated the code to use `INTERFACE_COMPILE_DEFINITIONS` I've also removed  `-D` prefix I think that this wasn't necessary.

**Review by taketwo — DISMISSED** [review]

**Comment by schweitzer** [comment]
Hello everyone, any updates on this ? Should I do something ?

**Comment by SergioRAgostinho** [comment]
You're good. We just need a second pair of eyes here to approve but the review backlog is considerable.

File: cmake/Modules/FindQhull.cmake
Comment (SergioRAgostinho) [inline cmake/Modules/FindQhull.cmake:9]: This comment needs to be preserved. You're still adapting the behavior of the find script based on `QHULL_USE_STATIC`

File: cmake/Modules/FindQhull.cmake
Comment (SergioRAgostinho) [inline cmake/Modules/FindQhull.cmake:119]: The language option is the same for both configurations. Can you just set `IMPORTED_LINK_INTERFACE_LANGUAGES` once?

File: cmake/Modules/FindQhull.cmake
Comment (schweitzer) [reply cmake/Modules/FindQhull.cmake:9]: Yep you're right, I've re-added the `QHULL_USE_STATIC` option.

File: cmake/Modules/FindQhull.cmake
Comment (schweitzer) [reply cmake/Modules/FindQhull.cmake]: Done

**Comment by taketwo** [comment]
@schweitzer could you please rebase on top of current `master` so that we can double-check if CI turns green?

@PointCloudLibrary/maintainers since this has some potential for breaking things in subtle ways, maybe merge after the release?

**Comment by SergioRAgostinho** [comment]
It was not included in the milestone so it was going to be out.

**Comment by schweitzer** [comment]
Hello everyone,

> @schweitzer could you please rebase on top of current `master` so that we can double-check if CI turns green?

I've rebased on latest master branch !

**Comment by larshg** [comment]
Yes, it seems to be the cullprint. I have installed qhull with the suggested PR they have at VCPKG and its seems to found fine.

However I don't get any printouts of the library locations and its found?

**Comment by larshg** [comment]
It finds qhull with the modern approach, ie. inside [these lines](https://github.com/PointCloudLibrary/pcl/pull/3741/files#diff-7b4d74bfb6818cfc9d3856ef5bd06383R45-R55).

Maybe print the qhull is found here - not sure if one can get the paths to libs as well somehow?

**Comment by larshg** [comment]
Hmm, it doesn't add the includes nor libs to the generated project. Still investigating.

**Comment by larshg** [comment]
Ahh, I had to add:

`set(HAVE_QHULL_2011 ON)` to get the correct includes.

and also the definitions:

```
add_definitions("-Dqh_QHpointer")
if(MSVC)
  add_definitions("-Dqh_QHpointer_dllimport")
endif()
```
for the DLL version. Using the above methods, though adds the definitions to all targets, maybe we can use some other methods to only add it to relevant projects.

Also the chaining of QHULL::QHULL -> Qhull::libqhull doesn't seem to work.

I had to manually add Qhull:libqhull to the surface project.

I'll have a look more tomorrow to verify everything again.

**Comment by larshg** [comment]
Is it true that ubuntu 16.04 is shipped with qhull 2015.2?
see : https://launchpad.net/ubuntu/xenial/+source/qhull

Then I guess we can remove the check if its qhull_2011, since its that or only newer from now on?

Then we can remove the HAVE_QHULL_2011 and the old includes?

**Comment by SergioRAgostinho** [comment]
> Is it true that ubuntu 16.04 is shipped with qhull 2015.2?
> see : https://launchpad.net/ubuntu/xenial/+source/qhull

You asked and answered your own question :)

> Then I guess we can remove the check if its qhull_2011, since its that or only newer from now on?

Yes. 16.04 is the lowest we support.

> Then we can remove the HAVE_QHULL_2011 and the old includes?

Yes.

**Comment by jspricke** [comment]
removing 2011 sounds ok to me: https://repology.org/project/qhull/versions

**Comment by larshg** [comment]
Yes, just wanted it to be verified as I'm not that used to read the package repositories 😄

**Comment by larshg** [comment]
I have opened a PR to @schweitzer PR, see https://github.com/IRCAD-IHU/pcl/pull/2.

It does however require, that https://github.com/microsoft/vcpkg/pull/9923 get merged first, for windows to be working, if installed using VCPKG ofc.

I guess if one install qhull in another way, its should work just fine.

**Comment by schweitzer** [comment]
> I have opened a PR to @schweitzer PR, see [IRCAD-IHU#2](https://github.com/IRCAD-IHU/pcl/pull/2).
> 
> It does however require, that [microsoft/vcpkg#9923](https://github.com/microsoft/vcpkg/pull/9923) get merged first, for windows to be working, if installed using VCPKG ofc.
> 
> I guess if one install qhull in another way, its should work just fine.

Thanks I'm gonna take a look and integrate your work in this branch asap.

File: cmake/Modules/FindQhull.cmake
Comment (larshg) [inline cmake/Modules/FindQhull.cmake:58]:
```suggestion
  endif()
  message(STATUS "Found QHull version ${Qhull_VERSION}")
```

**Comment by kunaltyagi** [comment]
Removing milestone due to upstream fix not arriving in time

**Comment by schweitzer** [comment]
Hello everybody,

Is there any news on this one ? Is it planned to merge this soon, or we still need to wait a fix from VCPKG ?

**Comment by larshg** [comment]
Well, not really. They just updated to 8.0.2 of qhull which is currently not found using master.

Im hoping to look into it soonish.

**Comment by larshg** [comment]
This one finds qhull 8.0.2 which is current version shipped on vcpkg. @kunaltyagi can you make the CI rerun to see if all is green?

**Comment by kunaltyagi** [comment]
I re-ran the CI, but I doubt it'd be green. The Windows CI is unlikely to work without changes to the boost version

**Comment by larshg** [comment]
Ahh, true. @schweitzer can you rebase on master?

**Comment by larshg** [comment]
All CI's does find qhull with this PR - so I think we should review this and get it merged.

**Comment by larshg** [comment]
I have added a PR towards @schweitzer PR.

**Comment by schweitzer** [comment]
@larshg I've merged your PR in this branch, and rebased onto latest master.

**Comment by larshg** [comment]
It seems that the old way with qh_qhPointer is not available anymore in Qhull 8.0.0 and therefore the parts using qhull will need a overhaul, to correctly use the new qhull_r or libqhullstatic libraries.
What we can do in this PR is to simple set QHull as not found, if the version is greater or equal than 8?

**Comment by schweitzer** [comment]
> It seems that the old way with qh_qhPointer is not available anymore in Qhull 8.0.0 and therefore the parts using qhull will need a overhaul, to correctly use the new qhull_r or libqhullstatic libraries.
> What we can do in this PR is to simple set QHull as not found, if the version is greater or equal than 8?
> 
> see http://www.qhull.org/news/qhull-news.html for reference.

I didn't check this PR for a while, sorry for that.

What about the QHull issue ? I can set QHull as not found for version greater or equal than 8, but is that ok to PCL maintainers ?

**Comment by larshg** [comment]
@schweitzer there has been a PR that now gives qhull 8 support, would you like to rebase this one and lets try to get ready for merge?

**Comment by schweitzer** [comment]
@larshg I've rebased this on latest master, I resolved some conflicts on the FindQHull.cmake, I think we should review it carefully.

Please let me know.

File: cmake/Modules/FindQhull.cmake
Comment (larshg) [inline cmake/Modules/FindQhull.cmake:35]: This one shouldn't have been removed?

**Comment by larshg** [comment]
@schweitzer could you solve issue I suggested?

**Pull request closed without merging**
